src/fsharp/PrettyNaming.fs

```diff
@@ -177,59 +269,89 @@ module internal Microsoft.FSharp.Compiler.PrettyNaming
-        (    cat = UnicodeCategory.UppercaseLetter // Letters
-          || cat = UnicodeCategory.LowercaseLetter 
-          || cat = UnicodeCategory.TitlecaseLetter
-          || cat = UnicodeCategory.ModifierLetter
-          || cat = UnicodeCategory.OtherLetter
-          || cat = UnicodeCategory.LetterNumber 
-        )
+        if c = '_' then true
+        else
+            match System.Char.GetUnicodeCategory c with
+            // Letters
+            | UnicodeCategory.UppercaseLetter
+            | UnicodeCategory.LowercaseLetter
+            | UnicodeCategory.TitlecaseLetter
+            | UnicodeCategory.ModifierLetter
+            | UnicodeCategory.OtherLetter
+            | UnicodeCategory.LetterNumber -> true
+            | _ -> false
 
     /// The characters that are allowed to be in an identifier.
     let IsIdentifierPartCharacter c =
-        let cat = System.Char.GetUnicodeCategory(c)
-        (    cat = UnicodeCategory.UppercaseLetter // Letters
-          || cat = UnicodeCategory.LowercaseLetter 
-          || cat = UnicodeCategory.TitlecaseLetter
-          || cat = UnicodeCategory.ModifierLetter
-          || cat = UnicodeCategory.OtherLetter
-          || cat = UnicodeCategory.LetterNumber 
-          || cat = UnicodeCategory.DecimalDigitNumber // Numbers
-          || cat = UnicodeCategory.ConnectorPunctuation // Connectors
-          || cat = UnicodeCategory.NonSpacingMark // Combiners
-          || cat = UnicodeCategory.SpacingCombiningMark
-          || c = '\'' // Tick
-        )
-
-    /// Is this character a part of a long identifier 
-    let IsLongIdentifierPartCharacter c = 
-        (IsIdentifierPartCharacter c) || (c = '.')
-
-    let IsValidPrefixOperatorUse s = 
+        if c = '\'' then true   // Tick
+        else
+            match System.Char.GetUnicodeCategory c with
+            // Letters
+            | UnicodeCategory.UppercaseLetter
+            | UnicodeCategory.LowercaseLetter
+            | UnicodeCategory.TitlecaseLetter
+            | UnicodeCategory.ModifierLetter
+            | UnicodeCategory.OtherLetter
+            | UnicodeCategory.LetterNumber
+            // Numbers
+            | UnicodeCategory.DecimalDigitNumber
+            // Connectors
+            | UnicodeCategory.ConnectorPunctuation
+            // Combiners
+            | UnicodeCategory.NonSpacingMark
+            | UnicodeCategory.SpacingCombiningMark -> true
+            | _ -> false
+
+    /// Is this character a part of a long identifier?
+    let IsLongIdentifierPartCharacter c =
+        c = '.'
+        || IsIdentifierPartCharacter c
+
+    let IsValidPrefixOperatorUse s =
+        if System.String.IsNullOrEmpty s then false else
         match s with 
         | "?+" | "?-" | "+" | "-" | "+." | "-." | "%" | "%%" | "&" | "&&" -> true
-        | _ -> s.[0] = '!' || (s.[0] = '~' && String.forall (fun c -> c = s.[0]) s)
+        | _ ->
+            s.[0] = '!'
+            || (String.forall (fun c -> c = '~') s)
```

Comment:
I think the first ``s.[0] = '~'``  is voiding a call to String.forall, just leave it as it was for this one